pkg/ocihook/ocihook.go

```diff
@@ -403,10 +422,15 @@ func onCreateRuntime(opts *handlerOpts) error {
 			ExtraHosts: opts.extraHosts,
 			Name:       opts.state.Annotations[labels.Name],
 		}
+		runtime.LockOSThread()
+		// nsents verified here we are in detached netwoprk ns
+		// nsPath verified is pointing to the nested detached ns
+		// user ns is the detch user ns
 		cniRes, err := opts.cni.Setup(ctx, opts.fullID, nsPath, namespaceOpts...)
```

Comment:
```
 time="2023-10-29T13:30:10Z" level=fatal msg="failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time=\"2023-10-29T13:30:10Z\" level=fatal msg=\"failed to call cni.Setup: plugin type=\\\"firewall\\\" failed (add): failed to list iptables chains: running [/usr/sbin/iptables -t filter -S --wait]: exit status 3: iptables v1.8.4 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)\\nPerhaps iptables or your kernel needs to be upgraded.\\n\"\nFailed to write to log, write /home/rootless/.local/share/nerdctl/1935db59/containers/nerdctl-test/0bb9ad1b47ae32ef3ffda72250edcf338ccbadf8d9572d9503e67faded379d57/oci-hook.createRuntime.log: file already closed: unknown"
```
https://github.com/containerd/nerdctl/actions/runs/6683660354/job/18160087021?pr=2535

Reply:
Can you dump `/proc/self/status` right after `netns.Set` to see if `CAP_NET_ADMIN` is gained in the namespaces?

Reply:
Maybe we need to re-exec the ocihook process to gain the caps

Reply:
Relevant:
- https://github.com/rootless-containers/rootlesskit/blob/v2.0.0-alpha.1/pkg/child/child.go#L257-L291

Reply:
> Can you dump `/proc/self/status` right after `netns.Set` to see if `CAP_NET_ADMIN` is gained in the namespaces?

```
SigPnd: 0000000000000000
ShdPnd: 0000000000000000
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: fffffffd7fc1feff
CapInh: 0000000000000000
CapPrm: 000001ffffffffff
CapEff: 000001ffffffffff
CapBnd: 000001ffffffffff
CapAmb: 0000000000000000
```

```
 capsh --decode=000001ffffffffff
0x000001ffffffffff=cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read,cap_perfmon,cap_bpf,cap_checkpoint_restore
```
`cap_net_admin` is already present right after `netns.Set`

Reply:
```
	if rootlessutil.IsRootlessChild() && detachNetNs {
		stateDir, err := rootlessutil.RootlessKitStateDir()
		if err != nil {
			return err
		}
		ns, err := netns.GetFromPath(filepath.Join(stateDir, "netns"))
		if err != nil {
			return err
		}
		if err = netns.Set(ns); err != nil {
			return fmt.Errorf("switch to detached netns: %w", err)
		}
		ok, err := rootlessutil.HasCaps()
		if err != nil {
			return err
		}
		fmt.Println(ok)
```
the above code is printing `true`

Reply:
Some progress in:
- https://github.com/containerd/nerdctl/pull/2723

We can just run the hook with `nsenter -n/detached/netns -- nerdctl internal oci-hook ...` and call it a day.